middle_end/flambda2/simplify/inlining/inlining_report.ml

```diff
@@ -4,128 +4,660 @@
+         that the inlining history is buggy or that we are simplifying the same
+         exact term several times in a row) but it seems that it can be
+         currently reached. I should investigate it a bit further.
+
+         Misc.fatal_errorf error *)
+      Some next
+    | None, next -> next
+    | old, None -> old
+
+  let insert_or_update_call ?decision_with_context ~dbg ~callee
+      ~(cont : item Map.t -> item Map.t) t =
+    let key = dbg, Key.Call callee in
+    Map.update key
+      (function
+        | None ->
+          Some
+            (Call
+               { inlined = cont Map.empty;
+                 decision_with_context;
+                 uid =
+                   Uid.create
+                     ~compilation_unit:
+                       (Inlining_history.Absolute.compilation_unit callee)
+                     (Inlining_history.Absolute.path callee)
+               })
+        | Some (Call t) ->
+          let decision_with_context =
+            join_decision t.decision_with_context decision_with_context
+              ~error:"Each call can have at most one associated decision"
+          in
+
+          Some
+            (Call
+               { inlined = cont t.inlined; decision_with_context; uid = t.uid })
+        | Some (Fundecl _ | Scope _) ->
+          Misc.fatal_errorf "%s"
+            "A key of type call should be associated with an item of type call")
+      t
+
+  let insert_or_update_fundecl ?before_simplify ?after_simplify
+      ~compilation_unit ~dbg ~name ~path ~(cont : item Map.t -> item Map.t) t =
+    let key = dbg, Key.Fundecl name in
+    Map.update key
+      (function
+        | None ->
+          Some
+            (Fundecl
+               { body = cont Map.empty;
+                 before_simplify;
+                 after_simplify;
+                 uid = Uid.create ~compilation_unit path
+               })
+        | Some (Fundecl t) ->
+          let before_simplify =
+            join_decision t.before_simplify before_simplify
+              ~error:
+                "Each function can have at most one associated decision for \
+                 the before simplify pass."
+          in
+          let after_simplify =
+            join_decision t.after_simplify after_simplify
+              ~error:
+                "Each function can have at most one associated decision for \
+                 the after simplify pass."
+          in
+          Some
+            (Fundecl
+               { body = cont t.body;
+                 after_simplify;
+                 before_simplify;
+                 uid = t.uid
+               })
+        | Some (Call _ | Scope _) ->
+          Misc.fatal_errorf
+            "A key of type call should be associated with an item of type call")
+      t
+
+  (* Insert a decision into an inlining tree [t].
+
+     Only decisions rooted in [compilation_unit] are kept. *)
+  let insert ~compilation_unit t (decision : raw_decision) =
+    let rec aux ~cont path =
```

Comment:
I'm really not sure what `aux` does. The name `cont` does make sense here, though

Reply:
Renamed `aux` and `cont`

Reply:
Did the renaming get dropped in a rebase? The current version still says `aux` and `cont`

Reply:
Fixed